Fix coinbase vulnerability

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This introduces a fix for a coinbase vulnerability by ensure that the coinbase is flagged correctly and the amount is exactly as required. 
Refactor block and aggregate body checks

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Currently the coinbase is only checked on the amount added to the block. So in essence this is checked by checking that a block emits the correct emission reward.  But its possible to manipulate the block to make the coinbase transaction very small, and add a second "coinbase" that does not have the lock heighjt. This means the block aoumt is till correct and the chain has no inflation. But the coinbase is incorrect. 
This PR forces the coinbase to be the exact amount as it should be. It also requires that coinbase transactions not have an offset in their blinding factors. Out current code also does not add this. The use of the offset is only their so that you cannot link a kernel to a output. But in the case of the coinbase, both contain flags and the consensus code check that there is only 1 utxo and 1 kernel flagged as coinbase. So there is no hiding the tx kernel here.  

The refactor of the block checks to aggregated body has been done as those test only run on the aggregated body it self. This refactor has the benefit of allowing transactions to access those checks as well as make the checks easier to test. 

WARNING
This test has the potential to break consensus if there was any shenanigans in the chain in the past. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+    pub fn check_kernel_rules(&self, height: u64) -> Result<(), TransactionError> {
+        for kernel in self.kernels() {
+            if kernel.lock_height > height {
+                debug!(target: LOG_TARGET, "Kernel lock height was not reached: {}", kernel);
```
Breaking rules should be a WARN, and it'll make more sense in the log context to include the `height` as well.

SWvheerden replied:
Done

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+            if utxo.features.flags.contains(OutputFlags::COINBASE_OUTPUT) {
+                coinbase_counter += 1;
+                if utxo.features.maturity < (height + consensus_constants.coinbase_lock_height()) {
+                    warn!(target: LOG_TARGET, "Coinbase found with maturity set too low");
```
You could include some info to id the utxo too in a noisy log context.

SWvheerden replied:
done

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+                coinbase_counter += 1;
+                if utxo.features.maturity < (height + consensus_constants.coinbase_lock_height()) {
+                    warn!(target: LOG_TARGET, "Coinbase found with maturity set too low");
+                    return Err(TransactionError::InvalidCoinbase);
```
Maybe we should have more granular Coinbase errors. => CoinbaseNotMatured

SWvheerden replied:
done

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+                target: LOG_TARGET,
+                "{} coinbases found in body. Only a single coinbase is permitted.", coinbase_counter,
+            );
+            return Err(TransactionError::InvalidCoinbase);
```
More granular Coinbase errors => IncorrectCoinbaseCount

SWvheerden replied:
Done

CjS77 commented on base_layer/core/src/validation/block_validators.rs:
```diff
@@ -378,3 +387,49 @@ fn check_cut_through(block: &Block) -> Result<(), ValidationError> {
     Ok(())
 }
+
+fn check_coinbase_output_dont_check_amount(
```
This is only called from statelessValidator -- but you could just call check_coinbase_output, so this function seems totally unnecessary.

SWvheerden replied:
This test is here to let some tests pass. 
Running it normally seems a lot of tests use the check_coinbase_output tests, but they do not have the correct output. One test actually uses this as a method to get an error.

CjS77 commented on base_layer/core/src/validation/block_validators.rs:
```diff
@@ -238,14 +241,20 @@ fn check_block_weight(block: &Block, consensus_constants: &ConsensusConstants) -
+    let total_coinbase = rules.calculate_coinbase_and_fees(block);
     block
-        .check_coinbase_output(consensus_constants)
+        .check_coinbase_output(total_coinbase, &rules.consensus_constants(), factories)
```
```suggestion
        .check_coinbase_output(total_coinbase, rules.consensus_constants(), factories)
```

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+    pub fn check_coinbase_output(
+        &self,
+        reward: MicroTari,
+        consensus_constants: &ConsensusConstants,
```
Rather take lock_height as an argument here, then you can drop the namespace import of ConsensusConstants AND it would make testing easier since you wouldn't have to create a whole CC object to test this fn.

SWvheerden replied:
done

CjS77 commented on base_layer/core/src/transactions/aggregated_body.rs:
```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
+            );
+            return Err(TransactionError::InvalidCoinbase);
+        }
+        aggregated_body
```
`validate_internal_consistency` is a very expensive call and it doesn't look to be saving any lines of code -- what with constructing a whole new AggregateBody --since you only really want to check the CB amount is correct. The signature check & Range proof check is going to be called again at some point when the full body is validated.

So just calculate and check Is CBUtxo.commitment == CBExcess + (Sum(fees) + reward).H? 
Then the checks above can be dramatically simplified too, e.g. something like

`counter = self.kernels().map_and_filter(|k| if k.features.contains(KernelFeatures::COINBASE_KERNEL) {1} else {0}).sum()`

SWvheerden replied:
Done. I think. 
You still need to find and map out the coinbase utxo and kernel to use its values.

CjS77 commented on base_layer/core/src/validation/block_validators.rs:
```diff
@@ -190,7 +193,7 @@ impl StatelessValidation<Block> for MockStatelessBlockValidator {
     /// 1. Was cut through applied in the block?
     fn validate(&self, block: &Block) -> Result<(), ValidationError> {
-        check_coinbase_output(block, &self.rules.consensus_constants())?;
+        check_coinbase_output_dont_check_amount(block, self.rules.clone(), &self.factories)?;
```
You can just use the usual  check_coinbase_output fn here. Don't see why you're creating a new version

SWvheerden replied:
check comment above

stringhandler commented on base_layer/core/src/validation/block_validators.rs:
```diff
@@ -378,3 +387,44 @@ fn check_cut_through(block: &Block) -> Result<(), ValidationError> {
     Ok(())
 }
+
+fn check_coinbase_output_dont_check_amount(block: &Block, rules: ConsensusManager) -> Result<(), ValidationError> {
```
This method seems unnecessary, can it be removed?

stringhandler commented on base_layer/core/src/transactions/coinbase_builder.rs:
```diff
@@ -255,5 +255,82 @@ mod test {
+            tx.body
+                .validate_internal_consistency(&BlindingFactor::default(), block_reward, &factories),
+            Ok(())
+        );
```
For sanity, can we add a call to one of the validators that catches the inflation created here?

stringhandler reviewed (dismissed):
Looks good. I will pull and test it locally. I would have liked to see a unit test that mirrors the exact vulnerability, but I think if you create an issue to create the test later, I'd be happy with that. Ideally there should be at least one test for each of the errors `InvalidKernel`, `InvalidCoinbaseMaturity`, `InvalidCoinbaseCount`, `InputMaturity`. 

The test that you've added covers the case where the incorrect fee is added, which is ok, but the block in that test also inflates the supply, which would hopefully is caught by another validator. A test that shows this vulnerability more clearly may be: 
1. Create a coinbase, set its value to 0.
2. Create another UTXO (as you have done) with the correct emission and fees that should have been in the coinbase.
3. Verify that the test finds it.

SWvheerden committed:
Fix coinbase vulnerability

Refactor block and aggregate body checks
Co-authored-by: Cayle Sharrock <CjS77@users.noreply.github.com>
added unit test showing vulnerability 
Added tests to cover all possible coinbase vaiidation errors

SWvheerden commented:
I added more tests to cover every single coinbase validation error there is

stringhandler reviewed (approved):
Happy with the new tests. (Didn't run locally, but will merge if CI passes)
